src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,413 @@
+   */
+  getPrivateProfile(): UserModel {
+    const displayProfile = Object.keys(this.profile).reduce(
+      (acc, currKey) => ({
+        ...acc,
+        [currKey]: this.profile[currKey]?.value,
+      }),
+      {},
+    )
+    return getUserModel(displayProfile)
+  }
+
+  getFieldPrivacy(field: string): string {
+    const currentPrivacy = this.profile[field]?.privacy
+
+    return currentPrivacy || this.profileSettings[field].defaultPrivacy || 'public'
+  }
+
+  async validateProfile(profile: any): Promise<{ isValid: boolean, errors: {} }> {
+    if (!profile) {
+      return { isValid: false, errors: {} }
+    }
+    const fields = Object.keys(profile).filter(prop => UserStorage.indexableFields[prop])
+
+    const validatedFields = await Promise.all(
+      fields.map(async field => ({
+        field,
+        valid: await UserStorage.isValidValue(field, profile[field]?.value, true),
+      })),
+    )
+
+    const errors = validatedFields.reduce((accErrors, curr) => {
+      if (!curr.valid) {
+        accErrors[curr.field] = `Unavailable ${curr.field}`
+      }
+      return accErrors
+    }, {})
+
+    const isValid = validatedFields.every(elem => elem.valid)
+    logger.debug({ fields, validatedFields, errors, isValid, profile })
+
+    return { isValid, errors }
+  }
+
+  setProfileFieldPrivacy(field: string, privacy: FieldPrivacy): Promise<ACK> {
+    let value = this.getProfileFieldValue(field)
+    return this.setProfileField(field, value, privacy, true)
+  }
+
+  /**
+   * return public user profile by field value
+   * @param field - Profile field value (email, mobile or wallet address value)
+   * @returns {object} profile - { name, avatar }
+   */
+  async getUserProfile(field?: string): { name: string, avatar: string } {
+    const attr = isMobilePhone(field) ? 'mobile' : isEmail(field) ? 'email' : 'walletAddress'
+
+    const profile = await this._getPublicProfile(attr, field)
+    const { fullName, avatar } = profile
+    if (profile == null) {
+      logger.info(`getUserProfile by field <${field}> `)
+      return { name: undefined, avatar: undefined }
+    }
+
+    logger.info(`getUserProfile by field <${field}>`, { avatar, fullName })
+    if (!fullName) {
+      logger.info(`cannot get fullName from gun by field <${field}>`, { fullName })
+    }
+    return { name: fullName, avatar }
+  }
+
+  notifyProfileUpdates() {
+    over(this.subscribersProfileUpdates)(this.profile)
+  }
+
+  subscribeProfileUpdates(callback: any => void) {
+    this.subscribersProfileUpdates.push(callback)
+    if (this.profile) {
+      callback(this.profile)
+    }
+  }
+
+  unSubscribeProfileUpdates() {
+    this.subscribersProfileUpdates = []
+  }
+
+  /**
+   * Delete profile
+   * @returns {Promise<[Promise<void>, Promise<*>]>}
+   */
+  deleteProfile(): Promise<boolean> {
+    this.unSubscribeProfileUpdates()
+
+    return Promise.all([this.removeAvatar(), this.profiledb.deleteProfile()]).then(() => (this.profile = {}))
```

Comment:
1. use async await
2. move all palces where you change values  of this.profile into a helper function
3.  this helper function should notify subscribers  about changes to profile